src/algebra/cubic_discriminant.lean

```diff
@@ -0,0 +1,316 @@
+by rw [of_a_b_eq_zero ha hb, degree_linear hc]
+
+lemma degree_of_a_b_c_eq_zero (ha : P.a = 0) (hb : P.b = 0) (hc : P.c = 0) (hd : P.d ≠ 0) :
+  P.to_poly.degree = 0 :=
+by rw [of_a_b_c_eq_zero ha hb hc, degree_C hd]
+
+lemma degree_of_zero (ha : P.a = 0) (hb : P.b = 0) (hc : P.c = 0) (hd : P.d = 0) :
+  P.to_poly.degree = ⊥ :=
+by rw [of_zero ha hb hc hd, degree_zero]
+
+lemma leading_coeff (ha : P.a ≠ 0) : P.to_poly.leading_coeff = P.a := leading_coeff_cubic ha
+
+lemma leading_coeff_of_a_eq_zero (ha : P.a = 0) (hb : P.b ≠ 0) : P.to_poly.leading_coeff = P.b :=
+by rw [of_a_eq_zero ha, leading_coeff_quadratic hb]
+
+lemma leading_coeff_of_a_b_eq_zero (ha : P.a = 0) (hb : P.b = 0) (hc : P.c ≠ 0) :
+  P.to_poly.leading_coeff = P.c :=
+by rw [of_a_b_eq_zero ha hb, leading_coeff_linear hc]
+
+lemma leading_coeff_of_a_b_c_eq_zero (ha : P.a = 0) (hb : P.b = 0) (hc : P.c = 0) :
+  P.to_poly.leading_coeff = P.d :=
+by rw [of_a_b_c_eq_zero ha hb hc, leading_coeff_C]
+
+end degree
+
+/-! ### Map across a homomorphism -/
+
+section map
+
+variables [semiring S] {φ : R →+* S}
+
+/-- Map a cubic polynomial across a semiring homomorphism. -/
+def map (φ : R →+* S) (P : cubic R) : cubic S := ⟨φ P.a, φ P.b, φ P.c, φ P.d⟩
+
+lemma map_to_poly : (map φ P).to_poly = polynomial.map φ P.to_poly :=
+by simp only [map, to_poly, map_C, map_X, polynomial.map_add, polynomial.map_mul,
+              polynomial.map_pow]
+
+end map
+
+end basic
+
+section roots
+
+open multiset
+
+/-! ### Roots over an extension -/
+
+section extension
+
+variables {P : cubic R} [comm_ring R] [comm_ring S] {φ : R →+* S}
+
+/-- The roots of a cubic polynomial. -/
+def roots [is_domain R] (P : cubic R) : multiset R := P.to_poly.roots
+
+lemma map_roots [is_domain S] : (map φ P).roots = (polynomial.map φ P.to_poly).roots :=
+by rw [roots, map_to_poly]
+
+theorem mem_roots_iff [is_domain R] (h0 : P.to_poly ≠ 0) (x : R) :
+  x ∈ P.roots ↔ P.a * x ^ 3 + P.b * x ^ 2 + P.c * x + P.d = 0 :=
+begin
+  rw [roots, mem_roots h0, is_root, to_poly],
+  simp only [eval_C, eval_X, eval_add, eval_mul, eval_pow]
+end
+
+theorem card_roots_le [is_domain R] [decidable_eq R] : P.roots.to_finset.card ≤ 3 :=
+begin
+  apply le_trans (to_finset_card_le P.to_poly.roots),
+  by_cases hP : P.to_poly = 0,
+  { apply le_trans (card_roots' P.to_poly),
+    rw [hP],
+    exact zero_le 3 },
+  { rw [← with_bot.coe_le_coe],
+    exact le_trans (card_roots hP) (le_trans degree_cubic_le $ le_refl 3) }
+end
+
+end extension
+
+variables {P : cubic F} [field F] [field K] {φ : F →+* K} {x y z : K}
+
+/-! ### Roots over a splitting field -/
+
+section split
+
+theorem splits_iff_card_roots (ha : P.a ≠ 0) : splits φ P.to_poly ↔ (map φ P).roots.card = 3 :=
+begin
+  replace ha : (map φ P).a ≠ 0 := (ring_hom.map_ne_zero φ).mpr ha,
+  change splits ((ring_hom.id K).comp φ) P.to_poly ↔ card (map φ P).to_poly.roots = 3,
+  rw [← splits_map_iff, ← map_to_poly, splits_iff_card_roots,
```

Comment:
If possible, try to avoid relying too much on definitional equality.
```suggestion
  conv_lhs { rw [←ring_hom.id_comp φ] },
  rw [roots, ← splits_map_iff, ← map_to_poly, splits_iff_card_roots,
```